velox/row/CompactRow.cpp

```diff
@@ -17,6 +17,153 @@
 #include "velox/vector/FlatVector.h"
 
 namespace facebook::velox::row {
+namespace {
+constexpr size_t kSizeBytes = sizeof(int32_t);
+
+void writeInt32(char* buffer, int32_t n) {
+  memcpy(buffer, &n, sizeof(int32_t));
+}
+
+int32_t readInt32(const char* buffer) {
+  int32_t n;
+  memcpy(&n, buffer, sizeof(int32_t));
+  return n;
+}
+
+template <TypeKind kind>
+void serializeTyped(
+    const raw_vector<vector_size_t>& rows,
+    uint32_t childIdx,
+    DecodedVector& decoded,
+    size_t valueBytes,
+    const raw_vector<uint8_t*>& nulls,
+    char* buffer,
+    std::vector<size_t>& offsets) {
+  const auto mayHaveNulls = decoded.mayHaveNulls();
+  const auto* rawData = decoded.data<char>();
+
+  for (auto i = 0; i < rows.size(); ++i) {
+    if (mayHaveNulls && decoded.isNullAt(rows[i])) {
+      bits::setBit(nulls[i], childIdx, true);
+    } else {
+      // Write fixed-width value.
+      memcpy(
+          buffer + offsets[i],
+          rawData + decoded.index(rows[i]) * valueBytes,
+          valueBytes);
+    }
+    offsets[i] += valueBytes;
+  }
+}
+
+template <>
+void serializeTyped<TypeKind::UNKNOWN>(
+    const raw_vector<vector_size_t>& rows,
+    uint32_t childIdx,
+    DecodedVector& decoded,
+    size_t valueBytes,
+    const raw_vector<uint8_t*>& nulls,
+    char* buffer,
+    std::vector<size_t>& offsets) {
+  for (auto i = 0; i < rows.size(); ++i) {
+    bits::setBit(nulls[i], childIdx, true);
+  }
+}
+
+template <>
+void serializeTyped<TypeKind::BOOLEAN>(
+    const raw_vector<vector_size_t>& rows,
+    uint32_t childIdx,
+    DecodedVector& decoded,
+    size_t valueBytes,
+    const raw_vector<uint8_t*>& nulls,
+    char* buffer,
+    std::vector<size_t>& offsets) {
+  const auto mayHaveNulls = decoded.mayHaveNulls();
+
+  for (auto i = 0; i < rows.size(); ++i) {
+    if (mayHaveNulls && decoded.isNullAt(rows[i])) {
+      bits::setBit(nulls[i], childIdx, true);
+    } else {
+      // Write fixed-width value.
+      reinterpret_cast<bool*>(buffer)[offsets[i]] =
+          decoded.valueAt<bool>(rows[i]);
```

Comment:
It seems boolean values are more commonly to be accessed with 'bits' methods.
> Unlike other types, boolean values cannot be accessed individually via a bool pointer. Instead, you interpret the buffer as a contiguous array of unsigned 64-bit integers and access individual values using helper methods from the “bits” namespace.

https://facebookincubator.github.io/velox/develop/vectors.html#buffers

Reply:
The `valueAt` call to bool type matches this specialization https://github.com/facebookincubator/velox/blob/2abfadf23ec698fb9ded437b97901563ec34b51c/velox/vector/DecodedVector.h#L432-L435

Reply:
I referred to the `reinterpret_cast<bool*>(buffer)[offsets[i]]` process. Shall we change to bits method?

Reply:
> I referred to the `reinterpret_cast<bool*>(buffer)[offsets[i]]` process. Shall we change to bits method like 'bits::setBit'?

@rui-mo bool type is one byte in `CompactRow`, so we're setting one byte here. Is it necessary to specifically set one bit? https://facebookincubator.github.io/velox/develop/serde/compactrow.html

Reply:
I see. Thanks for the explain!